File: 541.反转字符串-ii.js
/*
 * @lc app=leetcode.cn id=541 lang=javascript
 *
 * [541] 反转字符串 II
 */

// @lc code=start
/**
 * @param {string} s
 * @param {number} k
 * @return {string}
 */
const reverse = (s, start, end) => {
  let i, j, temp;
  i = start;
  j = end;
  while (i < j) {
    temp = s[i];
    s.replace(s[i], s[j]);
    console.log(s);
    s = s.replace(s[j], temp);
    i++;
    j--;
  }
  return s;
};
var reverseStr = function (s, k) {
  let i,
    len = s.length;
  for (i = 0; i < s.length; i += 2 * k) {
    if (len < k) s = reverse(s, i, s.length);
    else s = reverse(s, i, i + k - 1);
    len = len - 2 * k;
  }
  return s;
};
// @lc code=end

File: store/541.反转字符串-ii.c
/*
 * @lc app=leetcode.cn id=541 lang=c
 *
 * [541] 反转字符串 II
 */

// @lc code=start

void reverse(char *s, int start, int end)
{
  if (start >= end)
    return;
  int i, j;
  char temp;
  i = start;
  j = end;
  while (i < j)
  {
    temp = s[i];
    s[i] = s[j];
    s[j] = temp;
    i++;
    j--;
  }
}

char *reverseStr(char *s, int k)
{
  int start, end, i, remain, len = 0;
  while (s[len] != '\0')
    len++;

  i = 0;
  remain = len - i * (2 * k);
  start = 0;
  end = k - 1;

  while (i < len / (2 * k))
  {
    reverse(s, start, end);
    i++;
    remain = len - i * (2 * k);
    start = start + 2 * k;
    end = end + 2 * k;
  }

  if (remain < k)
    reverse(s, len - remain, len - 1);

  if (remain >= k && remain < 2 * k)
    reverse(s, len - remain, len - remain + k - 1);

  return s;
}
// @lc code=end
